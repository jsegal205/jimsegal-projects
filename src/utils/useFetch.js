import { useState, useEffect } from "react";
import axios from "axios";

// custom hook for performing GET request
const useFetch = (url, initialValue) => {
  const [data, setData] = useState(initialValue);
  const [loading, setLoading] = useState(true);
  useEffect(() => {
    const fetchData = async () => {
      try {
        setLoading(true);
        const response = await axios.get(url);
        setData(response.data);
      } catch ({ response }) {
        setData({ status: response.status });
      } finally {
        setLoading(false);
      }
    };
    fetchData();
  }, [url]);
  return { loading, data };
};

export default useFetch;
