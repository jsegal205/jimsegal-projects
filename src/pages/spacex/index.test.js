import React from "react";
import { BrowserRouter as Router } from "react-router-dom";

import useFetch from "../../utils/useFetch";
import SpaceX from "./index";
import { render } from "@testing-library/react";
import { toMatchDiffSnapshot } from "snapshot-diff";

expect.extend({ toMatchDiffSnapshot });

jest.mock("../../utils/useFetch", () => jest.fn());

describe("SpaceX Component", () => {
  describe("when `useFetch` is awaiting promise to resolve", () => {
    it("displays `Loading...`", () => {
      useFetch.mockReturnValue({
        loading: true,
        data: [],
      });

      const { container } = render(
        <Router>
          <SpaceX />
        </Router>
      );
      expect(container).toMatchSnapshot();
    });
  });

  describe("when `useFetch` returns data", () => {
    it("displays data", () => {
      useFetch.mockReturnValue({
        loading: false,
        data: [
          {
            mission_name: "mission_name",
            details: "details",
            launch_date_utc: 1,
            rocket: { rocket_name: "rocket_name" },
            launch_site: { site_name_long: "site_name_long" },
          },
        ],
      });

      const { container } = render(
        <Router>
          <SpaceX />
        </Router>
      );
      expect(container).toMatchSnapshot();
    });

    describe("when no data returned", () => {
      it("displays no data", () => {
        useFetch.mockReturnValue({
          loading: false,
          data: [],
        });

        const { container } = render(
          <Router>
            <SpaceX />
          </Router>
        );
        expect(container).toMatchSnapshot();
      });
    });

    describe("when 500 status returned", () => {
      it("displays error messaging", () => {
        useFetch.mockReturnValue({
          loading: false,
          data: {
            error: "server likely on fire",
            status: 500,
          },
        });

        const { container } = render(
          <Router>
            <SpaceX />
          </Router>
        );
        expect(container).toMatchSnapshot();
      });
    });
  });
});
