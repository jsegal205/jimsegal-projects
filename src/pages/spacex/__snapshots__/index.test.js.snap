// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SpaceX Component when \`useFetch\` is awaiting promise to resolve displays \`Loading...\` 1`] = `
<div>
  <section
    class="spacex-container"
  >
    <h2>
      Next Space X Launch
    </h2>
    <section
      class="loader"
    >
      <div
        class="loader-wheel"
      />
      <div
        class="loader-text"
      >
        Loading
      </div>
    </section>
    0
  </section>
</div>
`;

exports[`SpaceX Component when \`useFetch\` returns data displays data 1`] = `
<div>
  <section
    class="spacex-container"
  >
    <h2>
      Next Space X Launch
    </h2>
    <div
      class="spacex-mega"
    >
      <div>
        357674
         
        days
         
      </div>
      <div>
        10
         
        hours
         
      </div>
      <div>
        21
         
        minutes
         
      </div>
      <div>
        47
         
        seconds
         
      </div>
    </div>
    <h3>
      mission_name_future
    </h3>
    <div
      class="spacex-item"
    >
      <label>
        When:
      </label>
      <span>
        Wednesday, January 1, 3000, 12:00 AM
      </span>
    </div>
    <div
      class="spacex-item"
    >
      <label>
        Rocket:
      </label>
      <span>
        rocket_name_future
      </span>
    </div>
    <div
      class="spacex-item"
    >
      <label>
        Launch Site:
      </label>
      <span>
        site_name_long_future
      </span>
    </div>
    <div
      class="spacex-details"
    >
      future details
    </div>
  </section>
</div>
`;

exports[`SpaceX Component when \`useFetch\` returns data when 500 status returned displays error messaging 1`] = `
<div>
  <section
    class="spacex-container"
  >
    <h2>
      Next Space X Launch
    </h2>
    <section>
      <h3>
        Whoops! There was a problem loading
         
        SpaceX data.
      </h3>
      <p>
        Please reload browser to try again in a little bit.
      </p>
    </section>
  </section>
</div>
`;

exports[`SpaceX Component when \`useFetch\` returns data when no data returned displays no data 1`] = `
<div>
  <section
    class="spacex-container"
  >
    <h2>
      Next Space X Launch
    </h2>
    0
  </section>
</div>
`;
